protocol/parser/parser.go

```diff
@@ -163,6 +237,118 @@ func parse(rpcInput RPCInput, blockParser spectypes.BlockParser, dataSource int)
 	return retval, nil
 }
 
+type ParsedInput struct {
+	parsedBlock  int64
+	parsedHashes []string
+}
+
+func NewParsedInput() *ParsedInput {
+	return &ParsedInput{
+		parsedBlock:  spectypes.NOT_APPLICABLE,
+		parsedHashes: make([]string, 0),
+	}
+}
+
+func (p *ParsedInput) SetBlock(block int64) {
+	p.parsedBlock = block
+}
+
+func (p *ParsedInput) GetBlock() int64 {
+	return p.parsedBlock
+}
+
+func (p *ParsedInput) GetBlockHashes() ([]string, error) {
+	if len(p.parsedHashes) == 0 {
+		return nil, fmt.Errorf("no parsed hashes found")
+	}
+	return p.parsedHashes, nil
+}
+
+func ParseWithGenericParsers(rpcInput RPCInput, genericParsers []spectypes.GenericParser) (*ParsedInput, error) {
+	if len(genericParsers) == 0 {
+		return nil, fmt.Errorf("no generic parsers to use")
+	}
+
+	params := rpcInput.GetParams()
```

Comment:
you are assuming generic parsers always parse params, but they may also parse result, i think the syntax should be .params.x for params and .result.y for result, do not assume always params